File: src/002/java/002.java
import java.util.Map;
import java.util.HashMap;

public class fibsum {

  Map<Integer, Integer> memoize = new HashMap<>();

  public static void main(String... args) {
    fibsum fs = new fibsum();

    int i = 0, f = 0, sum = 0;
    do {
      f = fs.fib(i);
      if (f % 2 == 0) {
        sum += f;
      }
      i++;
    } while (f < 4000000);

    System.out.println(sum);

  }

  int fib(int n) {
    if (n == 0 || n == 1) {
      return 1;
    }
    else {
      if (memoize.containsKey(n)) {
        return memoize.get(n);
      } else {
        memoize.put(n, fib(n - 1) + fib(n - 2));
        return memoize.get(n);
      }
    }
  }
}

File: src/002/ruby/002.rb
class Fibsum
  attr_accessor :mfib

  def initialize()
    self.mfib = {}
    (self.mfib[0], self.mfib[1]) = [1, 1]
  end

  def fib(n)
    mfib[n] ||= begin
     mfib[n] = self.fib(n-1) + self.fib(n-2)
    end
  end

end

fs = Fibsum.new
sum, i = 0, 0
loop do
  f = fs.fib(i)
  sum += f if f.even?
  break if f > 4000000
  i += 1
end

puts sum
